### PR TITLE
Squiz/PostStatementComment: use a different error code for annotations

### LINKED ISSUE #560: Squiz.Commenting.PostStatementComment should allow trailing annotations from popular PHP tools
## Is your feature request related to a problem?
Currently, the `Squiz.Commenting.PostStatementComment` sniff makes it impossible to use trailing annotations like PHPUnit's `// @codeCoverageIgnore`. This happens because the sniff forbids inline comments after statements (with some exceptions).

## Describe the solution you'd like
The `Squiz.Commenting.PostStatementComment` already makes some exceptions and allows inline comments after statements in a few cases. Most notably, it allows PHPCS's `// phpcs:ignore` trailing annotation. I believe the sniff should also allow trailing annotations used by other popular PHP tools.

Here is a non-exhaustive list of trailing annotations that the sniff should allow:

- PHPUnit's `// @codeCoverageIgnore` (https://docs.phpunit.de/en/11.2/code-coverage.html#ignoring-code-blocks)
- PHPStan's `// @phpstan-ignore` and `// @phpstan-ignore-line` (https://phpstan.org/user-guide/ignoring-errors)

If this change is implemented, the code examples below, which currently are flagged as errors by the sniff, will stop being flagged:

```php
$a = 1; // @codeCoverageIgnore
$b = 2; // @phpstan-ignore-line
$c = 3; // @phpstan-ignore variable.undefined
```

Open questions:

- Are there other trailing annotations used by popular PHP tools that the sniff should allow as well?
- PHPUnit doesn't support anything after `// @codeCoverageIgnore` while PHPStan requires an error identifier and, optionally, a comment after `// @phpstan-ignore`. I think that the sniff doesn't need to conform to the exact rules used by each tool, and just checking that a given comment starts with the annotation supported by the tool is enough. So, for example, `$a = 1; // @codeCoverageIgnore some comment` is not supported by PHPUnit, but would not be flagged by the sniff.

## Additional context (optional)
This was originally discussed with @jrfnl in https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/525#pullrequestreview-2181764369.

If the change proposed here is accepted, this will allow PHPCS itself to use `// @codeCoverageIgnore` in its codebase to make PHPUnit ignore lines of code that cannot be covered by tests when generating the code coverage report. 

- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [x] I intend to create a pull request to implement this feature.

### PR DESCRIPTION
**Pull request by rodrigoprimo**
# Description

As discussed in #560, this PR changes `Squiz.Commenting.PostStatementComment` to use a different error code when an annotation is found. This will allow ruleset maintainers to selectively exclude the flagging of trailing annotations from their ruleset. For example, for PHPCS itself, this will make it possible to use the `// @codeCoverageIgnore` annotation where needed.

An annotation is defined as any comment that starts with an optional space, followed by a `@` and any character, as long as it is not a whitespace. For now, only `//` comments are supported as this is the only type of comment supported by this sniff.

This change only applies to PHP code as support for JS will be dropped soon, and JS doesn't seem to follow the same convention of annotations starting with `@`.

## Suggested changelog entry

`Squiz.Commenting.PostStatementComment`: use a different error code for annotations

## Related issues/external references

Fixes #560 

## Types of changes

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

### CODE REVIEW

File: src/Standards/Squiz/Sniffs/Commenting/PostStatementCommentSniff.php [hunk 1]
```diff
@@ -109,6 +109,14 @@ public function process(File $phpcsFile, $stackPtr)
             }
         }
 
+        if ($phpcsFile->tokenizerType === 'PHP'
+            && preg_match('|//\s?@[^\s]+|', $tokens[$stackPtr]['content']) === 1
```
Comment (jrfnl) [inline src/Standards/Squiz/Sniffs/Commenting/PostStatementCommentSniff.php:113]: This regex currently allows only for no or exactly one whitespace char. What is this based on ? Why would multiple spaces not be allowed ?

Also: should the `\s` (which includes horizontal space and new lines) be changed to the more specific `[ \t]` (only allow spaces or tabs) ?
Note: also means a test should be added with a tab between the `//` and the annotation.

File: src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc [hunk 2]
```diff
@@ -52,3 +52,7 @@ $match = match($foo // comment
 ) {
     1 => 1, // comment
 };
+
+$a = 1; // @codeCoverageIgnore
+$b = 2; //@phpstan-ignore variable.undefined
+$c = 3; //  @thisIsNotAnAnnotation - more than one space before the @ symbol.
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc:58]: Just realized, we're also still missing a test with a comment starting with an `@`, but without tag, so `// @ = add tag` (or something like that), which should not flagged as an annotation, but as a normal comment.

File: src/Standards/Squiz/Sniffs/Commenting/PostStatementCommentSniff.php (hunk 1 above)
Comment (rodrigoprimo) [reply src/Standards/Squiz/Sniffs/Commenting/PostStatementCommentSniff.php]: > This regex currently allows only for no or exactly one whitespace char. What is this based on ? Why would multiple spaces not be allowed ?

This was based on the PHPUnit behavior. It only recognizes `@codeCoverageIgnore` as an annotation if there is no or one space between `//` and the annotation. But after seeing your questions, I checked PHPStan and saw that it behaves differently. It accepts multiple spaces or tabs between `//` and the annotation (https://phpstan.org/r/5c00a06b-796b-4983-baca-64e93891f3ee). So, I went ahead and changed the regex to consider all variations supported by both PHPUnit and PHPStan.

> Also: should the \s (which includes horizontal space and new lines) be changed to the more specific [ \t] (only allow spaces or tabs) ?
Note: also means a test should be added with a tab between the // and the annotation.

Good point, I failed to consider this. I just updated the regex and added a test with a tab.

File: src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc (hunk 2 above)
Comment (rodrigoprimo) [reply src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc]: I just added those two tests.

**Comment by rodrigoprimo** [comment]
Thanks for your review, @jrfnl! I pushed a new commit to implement the changes you suggested and replied to your comments. Could you please take another look?

File: src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc [hunk 3]
```diff
@@ -52,3 +52,13 @@ $match = match($foo // comment
 ) {
     1 => 1, // comment
 };
+
+// Comments recognized as annotations by PHPCS.
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc:56]:
```suggestion
// Issue #560: Annotations should be reported separately and be non-auto-fixable as their meaning may change when moved.
```

File: src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc [hunk 4]
```diff
@@ -52,3 +52,13 @@ $match = match($foo // comment
...
+$a = 1; //@codeCoverageIgnore
+$b = 2; // @phpstan-ignore variable.undefined
+$c = 3; //  @phpstan-ignore variable.undefined
+$d = 4; //	@tabInsteadOfSpace
+
+// Comments that include `@`, but are not recognized as annotations by PHPCS.
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Commenting/PostStatementCommentUnitTest.inc:62]:
```suggestion
// Comments that include `@`, but are not recognized as annotations by this sniff.
```

**Review by jrfnl — APPROVED** [review]
@rodrigoprimo Thanks for the PR and fixing the regex (and updating the tests to match). Definitely better now.

Realized only when prepping for merge that the comments in the test case files could be made a little more descriptive. Left two suggestions in-line. Other than that all good.
If you like, I can fix up those comments when merging. Let me know.

**Comment by rodrigoprimo** [comment]
@jrfnl, I went ahead and committed your suggestions. Thanks.

**Pull request merged**